### PR TITLE
chore: git detach

### LINKED ISSUE #37439: [Task] Switch git disconnect method to super service
Notion Link: https://www.notion.so/Task-Switch-git-disconnect-method-to-super-service-13cfe271b0e2800ca4e3d9809ea51ef1

Notion Content:

### PR DESCRIPTION
**Pull request by sondermanish**
Fixes #37439 

## Automation

/ok-to-test tags="@tag.Git"

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
  - Introduced methods for committing artifacts and managing Git operations, including `commitArtifact`, `detachRemote`, and `publishArtifactPostCommit`.
  - Added functionality to list branches and check repository privacy.
  - Enhanced error handling for Git operations, improving reliability during commits and pushes.

- **Bug Fixes**
  - Updated parameter naming for consistency across methods, enhancing clarity.

- **Documentation**
  - Improved JavaDoc comments for better understanding of method purposes and parameters.

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/12158276745>
> Commit: f058fbe37f866ecee2cfb8fde30e4d4b62e2fd7f
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=12158276745&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Git`
> Spec:
> <hr>Wed, 04 Dec 2024 11:42:39 UTC

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 1]
```diff
@@ -852,13 +855,81 @@ private Mono<String> commitArtifact(
...
+
+                    if (isBaseGitMetadataInvalid(baseArtifact.getGitArtifactMetadata(), gitType)) {
+                        return Mono.error(new AppsmithException(
+                                AppsmithError.INVALID_GIT_CONFIGURATION,
+                                "Please reconfigure the artifact to connect to git repo"));
+                    }
+
+                    GitArtifactMetadata gitArtifactMetadata = baseArtifact.getGitArtifactMetadata();
+                    ArtifactJsonTransformationDTO jsonTransformationDTO = new ArtifactJsonTransformationDTO();
+                    jsonTransformationDTO.setRefType(RefType.BRANCH);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:887]: From param instead?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 2]
```diff
@@ -852,13 +855,81 @@ private Mono<String> commitArtifact(
...
+                .flatMap(tuple -> {
+                    List<String> localBranches = tuple.getT1();
+                    Artifact baseArtifact = tuple.getT2();
+
+                    baseArtifact.setGitArtifactMetadata(null);
+                    gitArtifactHelper.resetAttributeInBaseArtifact(baseArtifact);
+
+                    GitArtifactMetadata gitArtifactMetadata = baseArtifact.getGitArtifactMetadata();
+                    ArtifactJsonTransformationDTO jsonTransformationDTO = new ArtifactJsonTransformationDTO();
+                    jsonTransformationDTO.setRefType(RefType.BRANCH);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:906]: Here as well

File: app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java [hunk 3]
```diff
@@ -852,13 +855,81 @@ private Mono<String> commitArtifact(
...
+                    jsonTransformationDTO.setBaseArtifactId(gitArtifactMetadata.getDefaultArtifactId());
+                    jsonTransformationDTO.setRepoName(gitArtifactMetadata.getRepoName());
+                    jsonTransformationDTO.setArtifactType(baseArtifact.getArtifactType());
+                    jsonTransformationDTO.setRefName(gitArtifactMetadata.getBranchName());
+
+                    // Remove the parent application branch name from the list
+                    Mono<Boolean> removeRepoMono = gitHandlingService.removeRepository(jsonTransformationDTO);
+                    Mono<? extends Artifact> updatedArtifactMono = gitArtifactHelper.saveArtifact(baseArtifact);
+
+                    Flux<? extends Artifact> deleteAllBranchesFlux =
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:917]: Should we make this detached from the API so we stop failing like we currently do for larger number of branches? We can then batch this as well.

Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/git/central/CentralGitServiceCEImpl.java:917]: makes sense.

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**